src/types/weakref/reference.rs

```diff
@@ -0,0 +1,1200 @@
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PyWeakRef::new_with` will be replaced by `PyWeakRef::new_bound_with` in a future PyO3 version"
+        )
+    )]
+    pub fn new_with<T, C>(py: Python<'_>, object: T, callback: C) -> PyResult<&'_ PyWeakRef>
+    where
+        T: ToPyObject,
+        C: ToPyObject,
+    {
+        Self::new_bound_with(py, object, callback).map(Bound::into_gil_ref)
+    }
+
+    /// Constructs a new Weak Reference (`weakref.ref`/`weakref.ReferenceType`) for the given object with a callback.
+    ///
+    /// Returns a `TypeError` if `object` is not weak referenceable (Most native types and PyClasses without `weakref` flag) or if the `callback` is not callable or None.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::PyWeakRef;
+    ///
+    /// #[pyclass(weakref)]
+    /// struct Foo { /* fields omitted */ }
+    ///
+    /// #[pyfunction]
+    /// fn callback(wref: Bound<'_, PyWeakRef>) -> PyResult<()> {
+    ///         let py = wref.py();
+    ///         assert!(wref.upgrade::<Foo>()?.is_none());
+    ///         py.run_bound("counter = 1", None, None)
+    /// }
+    ///
+    /// # fn main() -> PyResult<()> {
+    /// Python::with_gil(|py| {
+    ///     py.run_bound("counter = 0", None, None)?;
+    ///     assert_eq!(py.eval_bound("counter", None, None)?.extract::<u32>()?, 0);
+    ///     let foo = Bound::new(py, Foo{})?;
+    ///
+    ///     // This is fine.
+    ///     let weakref = PyWeakRef::new_bound_with(py, foo.clone(), py.None())?;
+    ///     assert!(weakref.upgrade::<Foo>()?.is_some());
+    ///     assert!(
+    ///         // In normal situations where a direct `Bound<'py, Foo>` is required use `upgrade::<Foo>`
+    ///         weakref.get_object()
+    ///             .map_or(false, |obj| obj.is(&foo))
+    ///     );
+    ///     assert_eq!(py.eval_bound("counter", None, None)?.extract::<u32>()?, 0);
+    ///
+    ///     let weakref2 = PyWeakRef::new_bound_with(py, foo.clone(), wrap_pyfunction!(callback, py)?)?;
+    ///     assert!(!weakref.is(&weakref2)); // Not the same weakref
+    ///     assert!(weakref.eq(&weakref2)?);  // But Equal, since they point to the same object
+    ///
+    ///     drop(foo);
+    ///
+    ///     assert!(weakref.upgrade::<Foo>()?.is_none());
+    ///     assert_eq!(py.eval_bound("counter", None, None)?.extract::<u32>()?, 1);
+    ///     Ok(())
+    /// })
+    /// # }
+    /// ```
+    #[track_caller]
+    pub fn new_bound_with<T, C>(
+        py: Python<'_>,
+        object: T,
+        callback: C,
+    ) -> PyResult<Bound<'_, PyWeakRef>>
+    where
+        T: ToPyObject,
+        C: ToPyObject,
+    {
+        unsafe {
+            Bound::from_owned_ptr_or_err(
+                py,
+                ffi::PyWeakref_NewRef(
+                    object.to_object(py).as_ptr(),
+                    callback.to_object(py).as_ptr(),
```

Comment:
I didn't know about `PyCFunction::new_closure` yet.
However, wouldn't that limit the ability to use Python native functions, or callables?

Reply:
I think keeping it as an object reference (`Bound` or `Py`-GIL referenced) would make the most sense, since they would allow for the most flexibility while using the passing the callback (For example, functions defined in Python would also be usable).

Maybe an extra construct for taking in closures specifically would be nice. (e.g. `new_bound_with_closure`?)

That would probably result in the best of both worlds.